Sources/App/Core/SearchFilter.swift

```diff
@@ -0,0 +1,187 @@
+import Foundation
+import SQLKit
+import Vapor
+
+protocol SearchFilter {
+    // <Key>:<Op?><Value>
+    static var key: String { get }
+    
+    init(value: String, comparison: SearchFilterComparison) throws
+    func query(_ builder: SQLPredicateGroupBuilder) -> SQLPredicateGroupBuilder
+}
+
+struct SearchFilterParser {
+    
+    static var allSearchFilters: [SearchFilter.Type] = [
+        StarsSearchFilter.self,
+        LicenseSearchFilter.self,
+        UpdatedSearchFilter.self,
+    ]
+    
+    func separate(terms: [String]) -> (terms: [String], filters: [SearchFilter]) {
+        terms.reduce(into: (terms: [], filters: [])) { builder, term in
+            if let filter = parse(term: term) {
+                builder.filters.append(filter)
+            } else {
+                builder.terms.append(term)
+            }
+        }
+    }
+    
+    func parse(term: String) -> SearchFilter? {
+        let components = term
+            .components(separatedBy: ":")
+            .map { $0.trimmingCharacters(in: .whitespacesAndNewlines) }
+            .filter { !$0.isEmpty }
+        
+        guard components.count == 2 else {
+            return nil
+        }
+        
+        // Operator
+        let filterComparison: SearchFilterComparison = {
+            let comparisonOperator = String(components[1].prefix(1))
+            switch comparisonOperator {
+            case ">": return .greaterThan
+            case "<": return .lessThan
+            case "!": return .negativeMatch
+            default:  return .match
+            }
+        }()
+        
+        // Value
+        let stringValue = filterComparison == .match ? components[1] : String(components[1].dropFirst())
+        guard !stringValue.isEmpty else { return nil }
```

Comment:
This code works, but it's definitely not my fondest work. It is very much limited to single character operators preventing us from supporting `>=` etc. Probably something we could come back and rewrite later on without needing sweeping changes though.

Reply:
I definitely think you've touched on a few valid points here and in other comments, there will be extensions to the grammar here but only in ways that augments this starting point.

I think ultimately we'll want a system to allow strings with spaces, so that'll force us to tokenise speech marks correctly: `test:"some word"`

You mention the ability to check for iOS **and** macOS, this (currently) would just be written as `platform:ios platform:macos` since it's always using an "AND" clause. The complexity comes if you want an "OR" clause where we could do something like `framework:uikit|swiftui` (dumb example but you get the point). 

A lot to consider, but let's spin it out to a new issue? "Extend filter syntax grammar"